Improve the words sequence for lines

We should also consider the non-English words, such as for sentences like `find a place to eat 热干面` we should split it into:

- find
- a
- place
- to
- eat
- 热
- 干
- 面

The current way we do it is not performant:

https://github.com/ysmood/got/blob/2338a2006a4c77f126e418061f5628017351b690/lib/diff/sequence.go#L33-L36

> You can write some tests for it, I think yours should fail.

yeah actually I've tested it with

```
func TestNewString(t *testing.T) {
	out := diff.NewString("find a place to eat 热干面")
	fmt.Println(len(out))
	for _, c := range out {
		fmt.Println("String:",c," Hash:",c.Hash())
	}
}
```

and the result is 

```
=== RUN   TestNewString
23
String: f  Hash: f
String: i  Hash: i
String: n  Hash: n
String: d  Hash: d
String:    Hash:
String: a  Hash: a
String:    Hash:
String: p  Hash: p
String: l  Hash: l
String: a  Hash: a
String: c  Hash: c
String: e  Hash: e
String:    Hash:
String: t  Hash: t
String: o  Hash: o
String:    Hash:
String: e  Hash: e
String: a  Hash: a
String: t  Hash: t
String:    Hash:
String: 热  Hash: 热
String: 干  Hash: 干
String: 面  Hash: 面
--- PASS: TestNewString (0.00s)
PASS
```

Meanwhile, before I make changes, the test fails with panic

```
=== RUN   TestNewString
29
String: f  Hash: f
String: i  Hash: i
String: n  Hash: n
String: d  Hash: d
String:    Hash:
String: a  Hash: a
String:    Hash:
String: p  Hash: p
String: l  Hash: l
String: a  Hash: a
String: c  Hash: c
String: e  Hash: e
String:    Hash:
String: t  Hash: t
String: o  Hash: o
String:    Hash:
String: e  Hash: e
String: a  Hash: a
String: t  Hash: t
String:    Hash:
String: 热  Hash: 热
--- FAIL: TestNewString (0.00s)
panic: runtime error: invalid memory address or nil pointer dereference [recovered]
        panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x18 pc=0x5a2d2f]
```

That's why I asked about the purpose of this function.

--- a/lib/diff/sequence.go
+++ b/lib/diff/sequence.go
@@ -4,6 +4,7 @@ import (
 	"bufio"
 	"bytes"
 	"crypto/md5"
+	"unicode"
 )
 
 // Comparables list
@@ -17,24 +18,26 @@ type Comparable interface {
 	String() string
 }
 
-// Char is a rune
-type Char string
+// Word block
+type Word string
 
 // Hash interface
-func (c Char) Hash() string {
+func (c Word) Hash() string {
 	return string(c)
 }
 
 // String interface
-func (c Char) String() string {
+func (c Word) String() string {
 	return string(c)
 }
 
 // NewString from string
 func NewString(s string) Comparables {
-	cs := make([]Comparable, len(s))
-	for i, c := range s {
-		cs[i] = Char(c)
+	runes := []rune(s)
+	words := SplitSentence(runes)
+	cs := make([]Comparable, len(words))
+	for i, word := range words {
+		cs[i] = Word(word)
 	}
 	return cs
 }
@@ -78,4 +81,53 @@ func NewText(s string) Comparables {
 	}
 
 	return cs
+}
+
+// Split sentence into words, for word-wise comparing.
+// Any kind of splitter or non-English character is kept alone.
+func SplitSentence(sentence []rune) []string {
+	words := make([]string, 0)
+	var prev []rune
+	for _, r := range sentence {
+		if notPartOfWord(r) {
+			if len(prev) > 0 {
+				words = safelyAppendWord(words, string(prev))
+				prev = make([]rune, 0)
+			}
+			words = append(words, string(r))
+		} else {
+			prev = append(prev, r)
+		}
+	}
+	if len(prev) > 0 {
+		words = safelyAppendWord(words, string(prev))
+	}
+	return words
+}
+
+const MaxWordLen = 20
+
+// try not append very long word
+func safelyAppendWord(words []string, w string) []string {
+	if len(w) <= MaxWordLen {
+		return append(words, w)
+	}
+	parts := (len(w) + MaxWordLen - 1) / MaxWordLen
+	for i := 0; i < (parts - 1); i++ {
+		words = append(words, w[:MaxWordLen])
+		w = w[MaxWordLen:]
+	}
+	if len(w) > 0 {
+		words = append(words, w)
+	}
+	return words
+}
+
+// We assume that only alphabetic letters and digits can form word.
+func notPartOfWord(r rune) bool {
+	return !(inAlphabet(r) || unicode.IsDigit(r))
+}
+
+func inAlphabet(r rune) bool {
+	return (r >= 'a' && r <= 'z') || (r >= 'A' && r <= 'Z')
 }

--- a/lib/diff/sequence_test.go
+++ b/lib/diff/sequence_test.go
@@ -1,0 +1,35 @@
+package diff_test
+
+import (
+	"testing"
+
+	"github.com/ysmood/got/lib/diff"
+)
+
+func TestNewString(t *testing.T) {
+	g := setup(t)
+	testCases := []struct {
+		in   string
+		want []string
+	}{{
+		"find a place to eat 热干面",
+		[]string{"find", " ", "a", " ", "place", " ", "to", " ", "eat", " ", "热", "干", "面"},
+	}, {
+		"	as.Equal(arr, arr)",
+		[]string{"	", "as", ".", "Equal", "(", "arr", ",", " ", "arr", ")"},
+	}, {
+		"English-Words紧接着中文",
+		[]string{"English", "-", "Words", "紧", "接", "着", "中", "文"},
+	}, {
+		"WordVeryVeryVeryVeryVeryVeryVerylong",
+		[]string{"WordVeryVeryVeryVery", "VeryVeryVerylong"},
+	}}
+
+	for _, c := range testCases {
+		output := diff.NewString(c.in)
+		g.Eq(len(output), len(c.want))
+		for i, got := range output {
+			g.Eq(got.String(), c.want[i])
+		}
+	}
+}

